### PR TITLE
plumb through excess data gas

### PR DESCRIPTION
**Pull request by roberto-bayardo**
Post eip-4844, state transition processing requires access to the "excess data gas" value from the *parent* block's header. This requires a bit of new plumbing since it's the first field of this kind that needs to be available.

### CODE REVIEW

File: cmd/rpcdaemon/commands/eth_receipts.go [hunk 1]
```diff
@@ -56,10 +56,15 @@ func (api *BaseAPI) getReceipts(ctx context.Context, tx kv.Tx, chainConfig *para
 		}
 		return h
 	}
+	header := block.Header()
+	parentHeader := getHeader(header.ParentHash, header.Number.Uint64())
+	var excessDataGas *big.Int
+	if parentHeader != nil {
+		excessDataGas = parentHeader.ExcessDataGas
+	}
```
Comment (0x00101010) [inline cmd/rpcdaemon/commands/eth_receipts.go]: extract to helper function? saw it is duplicated in most functions

File: core/evm.go [hunk 2]
```diff
@@ -30,7 +30,10 @@ import (
 )
 
 // NewEVMBlockContext creates a new context for use in the EVM.
-func NewEVMBlockContext(header *types.Header, blockHashFunc func(n uint64) common.Hash, engine consensus.Engine, author *common.Address) vm.BlockContext {
+//
+// excessDataGas must be set to the excessDataGas value from the *parent* block header, and can be
+// nil if the parent block is not of EIP-4844 type. It is read only.
+func NewEVMBlockContext(header *types.Header, excessDataGas *big.Int, blockHashFunc func(n uint64) common.Hash, engine consensus.Engine, author *common.Address) vm.BlockContext {
```
Comment (0x00101010) [inline core/evm.go:36]: Maybe I missed something, but seems like we can get `excessDataGas` from `header` with below, am I missing something?

Same question for ApplyTransaction

```
        header := block.Header()
	parentHeader := getHeader(header.ParentHash, header.Number.Uint64())
	var excessDataGas *big.Int
	if parentHeader != nil {
		excessDataGas = parentHeader.ExcessDataGas
	}
```

File: cmd/rpcdaemon/commands/eth_receipts.go (hunk 1 above)
Comment (roberto-bayardo) [reply cmd/rpcdaemon/commands/eth_receipts.go]: good idea I think at least for the places where there is a "getHeader" func avalable we should hide behind a helper func. will tackle this tomorrow.

File: core/evm.go (hunk 2 above)
Comment (roberto-bayardo) [reply core/evm.go:36]: there is no "getHeader" func available here though.   For apply transaction you probably wouldn't want to be looking it up each time either, so passing in a cached value (you can reuse the same value for all txs in the same block) is going to be far more efficient.

File: cmd/rpcdaemon/commands/eth_receipts.go (hunk 1 above)
Comment (roberto-bayardo) [reply cmd/rpcdaemon/commands/eth_receipts.go]: done, ptal!

File: core/types/block.go [hunk 3]
```diff
@@ -103,6 +103,17 @@ type Header struct {
 	VerkleKeyVals []verkle.KeyValuePair
 }
 
+// ParentExcessDataGas is a helper that returns the excess data gas value of the parent block.  It
+// returns nil if the parent header could not be fetched, or if the parent block's excess data gas
+// is nil.
+func (h *Header) ParentExcessDataGas(getHeader func(hash common.Hash, number uint64) *Header) *big.Int {
```
Comment (0x00101010) [inline core/types/block.go:109]: 👍

**Review by 0x00101010 — APPROVED** [review]

**Pull request merged**